### ISSUE TITLE
node down ,statefulset  can't  create a replacement Pod with that same identity on the other nodes;

### ISSUE THREAD

**Issue by huangyanhong** [kubernetes/kubernetes#64869]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
> kind bug

**What happened**: I use k8s 1.8.4 
 I create statefulset by  https://www.kubernetes.org.cn/statefulset, create storage-class using the ceph rbd,then reboot the node running the statefulset, 
 the node can‘t shutdown 
showing:
blocking unmount of 
the /var/lib/kubelet/plugins/kubernetes.io/rbd/rbd/rbd-image-kubernetes-dynamic-pvc-d40dce8f-698a-11e8-9650-005056b751f5 type ext4 
the statefulset  pod change to r ‘Unknown’ states  5 minutes later ,but can’t create a replacement Pod with that same identity on the other nodes;

**What you expected to happen**:
when node down ,can create a replacement statefulset Pod with that same identity on the other nodes;

**How to reproduce it (as minimally and precisely as possible)**:
can't shutdown the node 

**Anything else we need to know?**:
If you find that some pods in your StatefulSet are stuck in the ‘Terminating’ or ‘Unknown’ states for an extended period of time, you may need to manually intervene to forcefully delete the pods from the apiserver?

**Environment**:
- Kubernetes version (use `kubectl version`): 1.8.4
- Cloud provider or hardware configuration: ceph 1.12.5
- OS (e.g. from /etc/os-release): centos 7.3
- Kernel (e.g. `uname -a`): 
3.10.0-693.el7.x86_64
- Install tools:
install ceph-common 1.12.5 every node
- Others:
use the storageclass

**Comment by tnozicka** [comment]
This is docummented here:
  https://kubernetes.io/docs/tasks/run-application/force-delete-stateful-set-pod/

TLDR: It is expected not to reschedule the statefulset pod while the node is down to avoid split brain / guarantee exactly once semantics. 

I have encountered this few times myself, not really happy about the fact that it can't auto recover from such state, especially since I had only 1 pod in the statefulset, but it has its reasons.

Maybe we could do a heartbeat/election from a side car to check in and have it kill the pod otherwise. Controller could then safely assume that if it din't renew its timestamp it can force delete the pod as it was self terminated even if the network split/node is down scenario. (leader election using the object field)

Might be better to come with system level support than the sidecar as this issue exists in all controllers running "named" pods.

/cc @smarterclayton @kubernetes/sig-apps-feature-requests
/remove-lifecycle stale

**Comment by smarterclayton** [comment]
System level support would be a fencing controller - there are several sets
of folks who are attempting that right now.



On Sep 6, 2018, at 4:52 AM, Tomáš Nožička <notifications@github.com> wrote:

This is docummented here:
https://kubernetes.io/docs/tasks/run-application/force-delete-stateful-set-pod/

TLDR: It is expected not to reschedule the statefulset pod while the node
is down to avoid split brain / guarantee exactly once semantics.

I have encountered this few times myself, not really happy about the fact
that it can't auto recover from such state, especially since I had only 1
pod in the statefulset, but it has its reasons.

Maybe we could do a heartbeat/election from a side car to check in and have
it kill the pod otherwise. Controller could then safely assume that if it
din't renew its timestamp it can force delete the pod as it was self
terminated even if the network split/node is down scenario. (leader
election using the object field)

Might be better to come with system level support than the sidecar as this
issue exists in all controllers running "named" pods.

/cc @smarterclayton <https://github.com/smarterclayton>
@kubernetes/sig-apps-feature-requests
<https://github.com/orgs/kubernetes/teams/sig-apps-feature-requests>
/remove-lifecycle stale

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/64869#issuecomment-419016450>,
or mute the thread

**Comment by tnozicka** [comment]
I think this is ok to close, given it's the intended behaviour and the system level feature is tracked elsewhere.

/close
